.ci/magician/cmd/generate_downstream.go

```diff
@@ -0,0 +1,338 @@
+			return "", err
+		}
+		scratchRepo.Branch = fmt.Sprintf("auto-pr-%s-old", ref)
+		commitMessage = fmt.Sprintf("Old generated code for MM PR %s.", ref)
+	case "downstream":
+		scratchRepo.Branch = "downstream-pr-" + ref
+		originalMessage, err := rnr.Run("git", []string{"log", "-1", "--pretty=%B", ref}, nil)
+		if err != nil {
+			return "", err
+		}
+		commitMessage = fmt.Sprintf("%s\n[upstream:%s]", originalMessage, ref)
+	}
+	return commitMessage, nil
+}
+
+func runMake(upstreamRepo *source.Repo, command string, rnr *exec.Runner) error {
+	switch upstreamRepo.Title {
+	case "terraform-google-conversion":
+		if _, err := rnr.Run("make", []string{"clean-tgc", "OUTPUT_PATH=" + upstreamRepo.Path}, nil); err != nil {
+			return err
+		}
+		if _, err := rnr.Run("make", []string{"tgc", "OUTPUT_PATH=" + upstreamRepo.Path}, nil); err != nil {
+			return err
+		}
+		if command == "downstream" {
+			if err := rnr.PushDir(upstreamRepo.Path); err != nil {
+				return err
+			}
+			if _, err := rnr.Run("go", []string{"get", "-d", "github.com/hashicorp/terraform-provider-google-beta@" + upstreamRepo.Branch}, nil); err != nil {
+				return err
+			}
+			if _, err := rnr.Run("go", []string{"mod", "tidy"}, nil); err != nil {
+				return err
+			}
+			if _, err := rnr.Run("make", []string{"build"}, nil); err != nil {
+				fmt.Println("Error building tgc: ", err)
+			}
+			if err := rnr.PopDir(); err != nil {
+				return err
+			}
+		}
+	case "tf-oics":
+		if _, err := rnr.Run("make", []string{"tf-oics", "OUTPUT_PATH=" + upstreamRepo.Path}, nil); err != nil {
+			return err
+		}
+	case "terraform":
+		if _, err := rnr.Run("make", []string{"clean-provider", "OUTPUT_PATH=" + upstreamRepo.Path}, nil); err != nil {
+			return err
+		}
+		if _, err := rnr.Run("make", []string{"provider", "OUTPUT_PATH=" + upstreamRepo.Path, fmt.Sprintf("VERSION=%s", upstreamRepo.Version)}, nil); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func pushCommit(upstreamRepo, scratchRepo *source.Repo, baseBranch, command, commitMessage, ref string, gh GithubClient, rnr *exec.Runner, ctlr *source.Controller) error {
+	if err := rnr.PushDir(scratchRepo.Path); err != nil {
+		return err
+	}
+	if err := setConfig(rnr); err != nil {
+		return err
+	}
+
+	if _, err := rnr.Run("git", []string{"add", "."}, nil); err != nil {
+		return err
+	}
+	if _, err := rnr.Run("git", []string{"checkout", "-b", scratchRepo.Branch}, nil); err != nil {
+		return err
+	}
+
+	if _, err := rnr.Run("git", []string{"commit", "--signoff", "-m", commitMessage}, nil); err != nil {
+		if strings.Contains(err.Error(), "nothing to commit, working tree clean") {
+			return nil
+		}
+		return err
+	}
+
+	prs, err := gh.GetPullRequests("closed", baseBranch, "updated", "desc")
+	if err != nil {
+		return err
+	}
+	var pullRequest github.PullRequest
+	for _, pr := range prs {
+		if pr.MergeCommitSha == ref {
+			pullRequest = pr
+			break
+		}
+	}
+
+	if command == "downstream" && upstreamRepo.Title == "terraform" {
```

Comment:
I believe this should be moved out of the push commit function to a new function called merge/push pull request and called from the toplevel `execGenerateDownstream`. That ways pushing commit and merging a PR are clear from the toplevel command.

Reply:
Done